### PR TITLE
Packing

### PR DESCRIPTION
**Pull request by andimarafioti**
Big boy refactor to make the balanced packing strategy from Nvidia's Eagle 2 paper in nanovlm. 

The good:
- Token throughput increased from 70k to 330k. 5-fold!
- Epoch duration reduced from 2900 seconds to 1300 seconds. 2.2 times! (I'm not sure where the difference with token throughput comes from, we now test on mmstar twice during one epoch, since epochs now have less steps, we used to do it more)
- GPU utilization is constantly high
- Validation loss looks great. Accuracy on MMStar is ok, it seems more stable than before, so the peaks aren't as high.

The PR has lots of commits since I went a bit back and forth on how to do it. The final result is:

Normal Datasets -> They now load images and text, prepare the prompts, apply chat templates, and tokenize the inputs. They also preprocess images (only resizing by now). They return a dict of labels, input_ids, attention_masks, processed_images.

Constant Length Dataset -> Takes one of the normal datasets and packs it. For this, it dynamically gets a large number of samples (approx 10 batches from before), and then used them to create groups of packed samples. Packed samples are simply one sample after another one from the normal datasets until reaching the max sequence length. The "balanced" part is that every group should have a similar distribution of lengths from samples. I took the main code for that from the Nvidia Eagle 2 paper. 
Because this processing takes ±3 mins, I create a queue and fill it in a different thread from the dataloader worker. That queue will fill up to 4 full batches such that the dataset always has samples to return. In practice, this is then mostly done while the GPU is busy with fwd/bwd, so we never wait for samples and GPU utilization is maxxed out.
Also, this constant length dataset manually shards itself when it detects several workers on the dataloader. This is needed since we can't use a default sampling strategy from PyTorch. The implementation directly divides the indices to avoid having each instance of the constant length dataset iterate over all datasets redoing the preprocessing.

Collators -> They are quite dumb, basically take samples from a dataset and discard samples that are too long and pad. If using with a constant-length dataset, there won't be samples that are too long, but I kept that so that it also works with the normal dataset. 

Training loop -> When creating the data, if we are on DDP, we manually shard the datasets. This is needed since we can't have a distributed sampler with an iterable dataset (like the constant length dataset). Then, because the datasets create packs and discards samples that are too long, not all ranks finish at the same time. So I added a wrapper on the dataloader to check if all the machines still have data, whenever one finished it's data, they all finish. This might drop a little bit of data, but nothing major. 
I also added timing logging to data loading, fwd/bwd, and post-processing. Useful to check speed-up improvements.

### CODE REVIEW

File: data/advanced_datasets.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        worker_id = worker_info.id if worker_info else 0
+        num_workers = worker_info.num_workers if worker_info else 1
+
+        def make_base_iterator():
+            """Return a (sharded) iterator over the underlying dataset."""
+            all_indices = range(len(self.dataset))
+
+            # Shard the *indices* first, before any data is fetched.
+            if num_workers > 1:
+                worker_indices = itertools.islice(all_indices, worker_id, None, num_workers)
```
[the remark this answers is not in the document]
Comment (andimarafioti) [reply data/advanced_datasets.py:40]: It doesn't materialize because it's still an iterator, just an iterator slice. Internally, this is wrapping the iterator in an if statement.

File: data/advanced_datasets.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                        iterator = make_iterator()
+                        self.epoch += 1
+                        print(f"Epoch {self.epoch} finished, restarting iterator")
+                        continue
+                    else:
+                        more_examples = False
+                        break
+
+                if len(sample["input_ids"]) > self.seq_length:
+                    continue  # skip overly long samples
```
[the remark this answers is not in the document]
Comment (andimarafioti) [reply data/advanced_datasets.py:90]: I can set seq_length to None per default, but we actually want this. Truncating or splitting samples in multimodal settings doesn't work (if text and other modalities aren't simultaneously attended to, you can't generate much)

File: data/advanced_datasets.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                        "labels": packed[1],
+                        "attention_mask": packed[2],
+                        "images": packed[3],
+                    }
+                )
+
+        # finished → unblock consumer
+        queue.put(self._sentinel)
+
+    def _balanced_greedy_knapsack(self, lengths, L, delta=0):
```
[the remark this answers is not in the document]
Comment (andimarafioti) [reply data/advanced_datasets.py:118]: This is a great idea! I'm taking this knapsack strategy from an NVidia paper (Eagle 2), and also thought that it was sad that images weren't balanced, but we can definitely add it 💡

**Comment by lusxvr** [comment]
The images are now balanced via long-tail cropping after a specified max amount right? And if a pack exceeds hits this image limit we pad the rest?

File: data/advanced_datasets.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andimarafioti) [reply data/advanced_datasets.py]: I tried this and talk to Kashif about it. Since images and text length are only losely correlated, this ends up being an open problem where you are optimizing for two variables. There is going to be a trade off. After looking at a few solutions, I ended up deciding to set a maximum number of images per knapsack, to avoid packs with too many images which would slow down training. This avoids me creating too many packs.

File: train.py [hunk 4]
```diff
@@ -100,7 +105,12 @@ def get_dataloaders(train_cfg, vlm_cfg):
     train_size = total_samples - val_size
 
     train_dataset = VQADataset(train_ds.select(range(train_size)), tokenizer, image_processor, vlm_cfg.mp_image_token_length)
+    
+    train_dataset = ConstantLengthDataset(train_dataset, infinite=False, max_sample_length=train_cfg.max_sample_length, seq_length=vlm_cfg.lm_max_length, num_of_sequences=train_cfg.batch_size*4, queue_size=train_cfg.batch_size*4*2,
```
Comment (lusxvr) [inline train.py:109]: This is missing in the config

Comment (andimarafioti) [reply train.py:109]: added it, sorry

**Comment by lusxvr** [comment]
I am still getting this warning from the tokenizer 
```
Training summary: 686170 samples, 42885 batches/epoch, batch size 64
Validation summary: 41801 samples, 2612 batches/epoch, batch size 64
Using device: cuda
Token indices sequence length is longer than the specified maximum sequence length for this model (8777 > 8192). Running this sequence through the model will result in indexing errors
Token indices sequence length is longer than the specified maximum sequence length for this model (11477 > 8192). Running this sequence through the model will result in indexing errors
Token indices sequence length is longer than the specified maximum sequence length for this model (8269 > 8192). Running this sequence through the model will result in indexing errors
Token indices sequence length is longer than the specified maximum sequence length for this model (8675 > 8192). Running this sequence through the model will result in indexing errors```

I thought this would disappear with the packing but we still need to tokenize every sample to see how long it would be right? Should we give the tokenizer the `max_lenght` as well? Or could this truncation interfere with the packing logic?

File: data/advanced_datasets.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                    continue  # skip samples that exceed the image constraint
+
+                sample["input_ids"] = torch.cat(
+                    [
+                        sample["input_ids"],
+                        torch.tensor([self.dataset.tokenizer.pad_token_id]),
+                    ]
+                )
+                sample["attention_mask"] = torch.cat(
+                    [sample["attention_mask"], torch.tensor([0])]
```
Comment (lusxvr) [inline data/advanced_datasets.py:130]: Shouldn't we attend to the added pad/eos token so that the model learns to generate an eos token when its finished with an answer?

Comment (lusxvr) [reply data/advanced_datasets.py:130]: Or is this simply a padding token so that the packed samples are differentiated?

File: train.py (hunk 4 above)
Comment (lusxvr) [inline train.py:109]: Should we set `infinite=train_cfg.epochs==1`?

Comment (lusxvr) [reply train.py:109]: My bad, I meant `infinite=train_cfg.epochs>1`. Or how is it handled if I train multiple epochs but have `infinite=False` for the dataset?

Comment (lusxvr) [reply train.py:109]: > Knowing how long the epochs are is obviously important for learning rate scheduling

True, does the old logic still hold? 

`max_steps = len(train_loader) * train_cfg.epochs // train_cfg.gradient_accumulation_steps`

Comment (andimarafioti) [reply train.py]: we chatted about this on slack
